app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/actions/AuthenticatedActionWithRetrievedData.scala

```diff
@@ -60,87 +63,127 @@ class AuthenticatedActionWithRetrievedData @Inject()(
+                  case cl ~ Some(AffinityGroup.Individual) ~ maybeNino ~ maybeSautr ~ maybeEmail ~ _ ~ _
+                      if cl < ConfidenceLevel.L200 =>
+                    Right(
+                      AuthenticatedRequestWithRetrievedData(
+                        UserType.IndividualWithInsufficientConfidenceLevel(
+                          maybeNino.map(NINO(_)),
+                          maybeSautr.map(SAUTR(_)),
+                          maybeEmail.filter(_.nonEmpty).map(Email(_)),
+                          ggCredId
+                        ),
+                        request
+                      )
+                    )
+
+                  case individual @ _ ~ Some(AffinityGroup.Individual) ~ _ ~ _ ~ _ ~ _ ~ _ =>
+                    individual match {
+                      case _ ~ _ ~ Some(nino) ~ _ ~ maybeEmail ~ _ ~ _ =>
+                        Right(
+                          AuthenticatedRequestWithRetrievedData(
+                            UserType.Individual(
+                              Right(NINO(nino)),
+                              maybeEmail.filter(_.nonEmpty).map(Email(_))
+                            ),
+                            request
+                          )
+                        )
+                    }
+
+                  case _ @_ ~ Some(AffinityGroup.Organisation) ~ _ ~ _ ~ maybeEmail ~ enrolments ~ _ =>
+                    handleOrganisation(request, enrolments, maybeEmail)
+
+                  case _ @_ ~ otherAffinityGroup ~ _ ~ _ ~ _ ~ _ ~ creds =>
+                    logger.warn(s"Got request for unsupported affinity group $otherAffinityGroup")
+                    Left(errorHandler.errorResult()(request))
+                }
+                result
               }
-
-            case _@_ ~ Some(AffinityGroup.Organisation) ~ _ ~ _ ~ maybeEmail ~ enrolments ~ _ =>
-              handleOrganisation(request, enrolments, maybeEmail)
-
-            case _@_ ~ otherAffinityGroup ~ _ ~ _ ~ _ ~ _ ~ creds =>
-              logger.warn(s"Got request for unsupported affinity group $otherAffinityGroup")
-              Left(errorHandler.errorResult()(request))
-
           }
-
-          Future.successful(result)
-        }
       }
-
   }
 
+  private def hasSubscribed[A](enrolments: Enrolments, request: MessagesRequest[A])(
+    implicit hc: HeaderCarrier
+  ): Future[Either[Result, Option[CgtReference]]] =
+    enrolments.getEnrolment("HMRC-CGT-PD") match {
+      case Some(cgtEnrolment) =>
+        cgtEnrolment.getIdentifier("CGTPDRef") match {
+          case Some(cgtReference) => Future.successful(Right(Some(CgtReference(cgtReference.value))))
+          case None               => Future.successful(Left(errorHandler.errorResult()(request)))
+        }
+      case None =>
+        subscriptionService.hasSubscription().value.map {
+          case Left(_) => Left(errorHandler.errorResult()(request))
+          case Right(maybeCgtReference) =>
+            maybeCgtReference match {
```

Comment:
do you need to pattern match on `maybeCgtReference`? Looks like you can probably just do a `leftMap` on the `hasSubscription` `EitherT` and then `.value` that

Reply:
also you could just do `fold` on the `EitherT` rather than doing `.value` and having `case` statements

Reply:
oh wait, this is equivalent to:
```scalasubscriptionService.hasSubscription().leftMap{
          e =>
          logger.warn("Could not check subscription status", e)
          errorHandler.errorResult()(request)
        }
```
(with an extra logging message)

Reply:
I prefer this:

``scala
      case None =>
        subscriptionService
          .hasSubscription()
          .leftMap(_ => errorHandler.errorResult()(request))
          .value
```